shared/src/metabase/shared/formatting/internal/numbers.clj

```diff
@@ -0,0 +1,109 @@
+                                            "E0"))
+          "currency"   (NumberFormat/getCurrencyInstance locale)
+          (doto (NumberFormat/getInstance locale)
+            (.setMaximumFractionDigits (:maximum-fraction-digits options 300))))
+        ;; Certain options do not map into Java's NumberFormat classes. They are handled by preprocessing
+        ;; the number (eg. by rounding) instead.
+        pre (if (:maximum-significant-digits options)
+              #(sig-figs % (:maximum-significant-digits options))
+              identity)]
+    ;; JavaScript does not support picking the rounding mode; it's always HALF_UP.
+    ;; (Intl.NumberFormat has an option `roundingMode` but it's new and not supported anywhere as of EOY2022.)
+    ;; Since Java is flexible, we match the HALF_UP behavior here.
+    (.setRoundingMode nf RoundingMode/HALF_UP)
+    (when (:minimum-fraction-digits options)
+      (.setMinimumFractionDigits nf (:minimum-fraction-digits options)))
+    (when currency
+      (.setCurrency nf (if (bad-currencies currency)
+                         ;; For the currencies the JVM doesn't support, we use USD and replace the symbols later.
+                         (Currency/getInstance "USD")
+                         (Currency/getInstance (name currency)))))
+    (when (:number-separators options)
+      (let [^DecimalFormat df nf
+            syms              (doto (.getDecimalFormatSymbols df)
+                                (.setDecimalSeparator  (first  (:number-separators options))))]
+        (if-let [grouping (second (:number-separators options))]
+          (.setGroupingSeparator syms grouping)
+          (.setGroupingUsed df false))
+        (.setDecimalFormatSymbols df syms)))
+    (reify
+      core/NumberFormatter
+      (format-number-basic [_ number]
+        (cond-> (.format nf (pre (bigdec (double number))))
+          ;; If running a "bad" currency Java doesn't support, replace the default symbol with the real one.
+          (and currency (bad-currencies currency))
+          (attach-currency-symbol nf locale currency)
+          ;; Handle the :currency-style option, which isn't supported natively on Java.
+          (and currency (not= (:currency-style options) "symbol"))
+          (apply-currency-style (.getCurrency nf) locale (:currency-style options) currency)))
+      (wrap-currency [_ text]
+        (let [sym (case currency
+                    :BTC "₿"
+                    (-> (name currency)
+                        (Currency/getInstance)
+                        (.getSymbol locale)))]
+          (str sym text)))
+      (split-exponent [_ formatted]
+        (let [^DecimalFormat df nf ;; Scientific mode always uses the DecimalFormat subclass.
+              exp (str/last-index-of formatted (.getExponentSeparator (.getDecimalFormatSymbols df)))]
+          {:mantissa (subs formatted 0 exp)
+           :exponent (subs formatted (inc exp))})))))
```

Comment:
I don't really love 70-line functions, they make my eyes bleed. It would be nice if it were possible to break out some of the different parts here into different subroutines. At the very least maybe you could do something like 

```clj
```clj
(defn- set-currency! [^NumberFormatter fomatter currency]
  ...)

(doto (number-formatter-for-style (:style options))
  ...
  (set-currency! (:currency options))
  ...)
```

Reply:
I've broken these up in both languages. They're still chunky, but now it's a few one-line `let` bindings and then the three methods under `reify`, which is hard to reduce further.